src/main/java/org/opensearch/knn/indices/ModelIndex.java

```diff
@@ -0,0 +1,229 @@
+        numberOfShards = MODEL_INDEX_NUMBER_OF_SHARDS_SETTING.get(settings);
+        numberOfReplicas = MODEL_INDEX_NUMBER_OF_REPLICAS_SETTING.get(settings);
+
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(MODEL_INDEX_NUMBER_OF_SHARDS_SETTING,
+                it -> numberOfShards = it);
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(MODEL_INDEX_NUMBER_OF_REPLICAS_SETTING,
+                it -> numberOfReplicas = it);
+    }
+
+    /**
+     * Creates model index. It is possible that the 2 threads call this function simulateously. In this case, one
+     * thread will throw a ResourceAlreadyExistsException. This should be caught and handled.
+     *
+     * @param actionListener CreateIndexResponse listener
+     * @throws IOException thrown when get mapping fails
+     */
+    public void createModelIndex(ActionListener<CreateIndexResponse> actionListener) throws IOException {
+        CreateIndexRequest request = new CreateIndexRequest(MODEL_INDEX_NAME)
+                .mapping("_doc", getMapping(), XContentType.JSON)
+                .settings(Settings.builder()
+                        .put("index.hidden", true)
+                        .put("index.number_of_shards", this.numberOfShards)
+                        .put("index.number_of_replicas", this.numberOfReplicas)
+                );
+        client.admin().indices().create(request, actionListener);
+    }
+
+    /**
+     * Checks if the model index exists
+     *
+     * @return true if the model index exists; false otherwise
+     */
+    public boolean modelIndexExists() {
+        return clusterService.state().getRoutingTable().hasIndex(MODEL_INDEX_NAME);
+    }
+
+    /**
+     * Put a model into the system index. Non-blocking
+     *
+     * @param modelId Id of model to create
+     * @param modelBlob byte array of model
+     * @param listener handles index response
+     */
+    public void putModel(String modelId, KNNEngine knnEngine, byte[] modelBlob, ActionListener<IndexResponse> listener) {
+        String base64Model = Base64.getEncoder().encodeToString(modelBlob);
+
+        Map<String, Object> parameters = ImmutableMap.of(
+                KNNConstants.KNN_ENGINE, knnEngine.getName(),
+                KNNConstants.MODEL_BLOB_PARAMETER, base64Model
+        );
+
+        IndexRequestBuilder indexRequestBuilder = client.prepareIndex(MODEL_INDEX_NAME, "_doc");
+        indexRequestBuilder.setId(modelId);
+        indexRequestBuilder.setSource(parameters);
+
+        putModel(indexRequestBuilder, listener);
+    }
+
+    /**
+     * Put a model into the system index. Non-blocking. When no id is passed in, OpenSearch will generate the id
+     * automatically. The id can be retrieved in the IndexResponse.
+     *
+     * @param modelBlob byte array of model
+     * @param listener handles index response
+     */
+    public void putModel(KNNEngine knnEngine, byte[] modelBlob, ActionListener<IndexResponse> listener) {
```

Comment:
is it possible to have different name ? method name overloading is not got for readability.

Reply:
I like method name overloading if the operation is the same. I think the parameters differentiate the 2 functions.

I could have `putWithModelId` and `putWithoutModelId` but I think this information is already encompassed in the parameters.